File: loopx/control_plane/agents/delivery_workspace.py
"""Python facade for the TypeScript-owned delivery workspace contract."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from ..coordination.coordination_state_contract_generated import (
    DELIVERY_WORKSPACE_SNAPSHOT_LEGACY_SNAPSHOT_SCHEMA,
    DELIVERY_WORKSPACE_SNAPSHOT_REQUEST_SCHEMA as DELIVERY_WORKSPACE_REQUEST_SCHEMA,
    DELIVERY_WORKSPACE_SNAPSHOT_RESULT_SCHEMA as DELIVERY_WORKSPACE_RESULT_SCHEMA,
    DELIVERY_WORKSPACE_SNAPSHOT_SNAPSHOT_SCHEMA as DELIVERY_WORKSPACE_SCHEMA_VERSION,
)
from ..effect_runtime import EffectRuntimeRejected, effect_runtime_result

LEGACY_DELIVERY_WORKSPACE_SCHEMA_VERSION = (
    DELIVERY_WORKSPACE_SNAPSHOT_LEGACY_SNAPSHOT_SCHEMA
)
DELIVERY_WORKSPACE_IDENTITY_KINDS = frozenset({"git_repository", "local_goal"})
DELIVERY_WORKSPACE_KINDS = frozenset(
    {"canonical_checkout", "independent_git_worktree", "local_goal_workspace"}
)


def _runtime_result(operation: str, **params: Any) -> Mapping[str, Any]:
    try:
        result = effect_runtime_result(
            "agent.delivery_workspace.evaluate",
            {
                "schema_version": DELIVERY_WORKSPACE_REQUEST_SCHEMA,
                "operation": operation,
                **params,
            },
        )
    except EffectRuntimeRejected as exc:
        raise RuntimeError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript delivery workspace result must be an object")
    if result.get("schema_version") != DELIVERY_WORKSPACE_RESULT_SCHEMA:
        raise RuntimeError("TypeScript delivery workspace result shape mismatch")
    return result


def _workspace_result(result: Mapping[str, Any]) -> dict[str, Any] | None:
    value = result.get("workspace")
    if value is None:
        return None
    if not isinstance(value, Mapping):
        raise RuntimeError("TypeScript delivery workspace result shape mismatch")
    expected_keys = {
        "schema_version",
        "workspace_identity",
        "identity_kind",
        "task_repository",
        "repository_source",
        "workspace_kind",
        "peer_independent_worktree_required",
    }
    actual_keys = set(value)
    if actual_keys not in (
        expected_keys,
        expected_keys | {"workspace_revision_digest"},
    ):
        raise RuntimeError("TypeScript delivery workspace result shape mismatch")
    if (
        value.get("schema_version") != DELIVERY_WORKSPACE_SCHEMA_VERSION
        or not isinstance(value.get("workspace_identity"), str)
        or value.get("identity_kind") not in DELIVERY_WORKSPACE_IDENTITY_KINDS
        or not isinstance(value.get("repository_source"), str)
        or value.get("workspace_kind") not in DELIVERY_WORKSPACE_KINDS
        or not isinstance(value.get("peer_independent_worktree_required"), bool)
        or (
            value.get("workspace_revision_digest") is not None
            and not isinstance(value.get("workspace_revision_digest"), str)
        )
        or (
            value.get("task_repository") is not None
            and not isinstance(value.get("task_repository"), str)
        )
    ):
        raise RuntimeError("TypeScript delivery workspace result shape mismatch")
    return dict(value)


def build_delivery_workspace_snapshot(
    *,
    workspace_identity: str,
    identity_kind: str,
    repository_source: str,
    workspace_kind: str,
    peer_independent_worktree_required: bool,
    workspace_revision_digest: str | None = None,
) -> dict[str, Any] | None:
    return _workspace_result(
        _runtime_result(
            "build",
            observation={
                "workspace_identity": workspace_identity,
                "identity_kind": identity_kind,
                "workspace_revision_digest": workspace_revision_digest,
                "repository_source": repository_source,
                "workspace_kind": workspace_kind,
                "peer_independent_worktree_required": bool(
                    peer_independent_worktree_required
                ),
            },
        )
    )


def normalize_delivery_workspace_snapshot(value: Any) -> dict[str, Any] | None:
    if not isinstance(value, Mapping):
        return None
    prepared = {
        "schema_version": value.get("schema_version"),
        "workspace_identity": value.get("workspace_identity"),
        "identity_kind": value.get("identity_kind"),
        "task_repository": value.get("task_repository"),
        "workspace_revision_digest": value.get("workspace_revision_digest"),
        "repository_source": value.get("repository_source"),
        "workspace_kind": value.get("workspace_kind"),
        "peer_independent_worktree_required": value.get(
            "peer_independent_worktree_required"
        ),
    }
    return _workspace_result(_runtime_result("normalize", workspace=prepared))

File: loopx/control_plane/agents/delivery_workspace.ts
import type { JsonObject } from "../effect_program.ts";
import { EffectRuntimeRequestError } from "../effect_runtime_errors.ts";
import {
  optionalNonEmptyString,
  requireBoolean,
  requireJsonObject,
  requireNonEmptyString,
  requireStringLiteral,
} from "../runtime_decode.ts";
import {
  DELIVERY_WORKSPACE_SNAPSHOT_LEGACY_SNAPSHOT_SCHEMA,
  DELIVERY_WORKSPACE_SNAPSHOT_REQUEST_SCHEMA,
  DELIVERY_WORKSPACE_SNAPSHOT_RESULT_SCHEMA,
  DELIVERY_WORKSPACE_SNAPSHOT_SNAPSHOT_SCHEMA,
} from "../coordination/coordination_state_contract.generated.ts";

export const DELIVERY_WORKSPACE_SCHEMA_VERSION =
  DELIVERY_WORKSPACE_SNAPSHOT_SNAPSHOT_SCHEMA;
export const LEGACY_DELIVERY_WORKSPACE_SCHEMA_VERSION =
  DELIVERY_WORKSPACE_SNAPSHOT_LEGACY_SNAPSHOT_SCHEMA;
export const DELIVERY_WORKSPACE_REQUEST_SCHEMA =
  DELIVERY_WORKSPACE_SNAPSHOT_REQUEST_SCHEMA;
export const DELIVERY_WORKSPACE_RESULT_SCHEMA =
  DELIVERY_WORKSPACE_SNAPSHOT_RESULT_SCHEMA;

export const DELIVERY_WORKSPACE_IDENTITY_KINDS = [
  "git_repository",
  "local_goal",
] as const;
export type DeliveryWorkspaceIdentityKind =
  (typeof DELIVERY_WORKSPACE_IDENTITY_KINDS)[number];

export const DELIVERY_WORKSPACE_KINDS = [
  "canonical_checkout",
  "independent_git_worktree",
  "local_goal_workspace",
] as const;
export type DeliveryWorkspaceKind =
  (typeof DELIVERY_WORKSPACE_KINDS)[number];

export interface DeliveryWorkspaceSnapshot extends JsonObject {
  schema_version: typeof DELIVERY_WORKSPACE_SCHEMA_VERSION;
  workspace_identity: string;
  identity_kind: DeliveryWorkspaceIdentityKind;
  task_repository: string | null;
  workspace_revision_digest?: string;
  repository_source: string;
  workspace_kind: DeliveryWorkspaceKind;
  peer_independent_worktree_required: boolean;
}

type DeliveryWorkspaceOperation = "build" | "normalize";

const GIT_IDENTITY_PATTERN =
  /^git:[a-z0-9.-]+(?::[0-9]{1,5})?\/[A-Za-z0-9._~+/-]+$/i;
const LOCAL_GOAL_IDENTITY_PATTERN =
  /^loopx:[A-Za-z0-9][A-Za-z0-9._:-]{0,255}$/;
const GIT_REVISION_DIGEST_PATTERN = /^[0-9a-f]{64}$/i;

function operation(value: unknown): DeliveryWorkspaceOperation {
  return requireStringLiteral(
    value,
    ["build", "normalize"] as const,
    "delivery workspace operation",
    "delivery workspace operation is unsupported",
  );
}

function requestObject(value: unknown): JsonObject {
  const request = requireJsonObject(value, "delivery workspace request");
  if (request.schema_version !== DELIVERY_WORKSPACE_REQUEST_SCHEMA) {
    throw new EffectRuntimeRequestError("delivery workspace request schema mismatch");
  }
  return request;
}

function canonicalGitIdentity(value: unknown, label: string): string | null {
  const identity = optionalNonEmptyString(value, label);
  if (!identity || !GIT_IDENTITY_PATTERN.test(identity)) return null;
  const suffix = identity.slice("git:".length);
  const slash = suffix.indexOf("/");
  if (slash <= 0) return null;
  const host = suffix.slice(0, slash).toLowerCase();
  let path = suffix.slice(slash + 1).replace(/\/{2,}/g, "/");
  if (path.endsWith(".git")) path = path.slice(0, -4);
  if (
    !path || path.split("/").some((segment) => segment === "." || segment === "..")
  ) return null;
  return `git:${host}/${path}`;
}

function localGoalIdentity(value: unknown, label: string): string | null {
  const identity = optionalNonEmptyString(value, label);
  return identity && LOCAL_GOAL_IDENTITY_PATTERN.test(identity) ? identity : null;
}

function snapshot(
  workspaceIdentity: string,
  identityKind: DeliveryWorkspaceIdentityKind,
  workspaceRevisionDigest: string | null,
  repositorySource: string,
  workspaceKind: DeliveryWorkspaceKind,
  peerIndependentWorktreeRequired: boolean,
): DeliveryWorkspaceSnapshot | null {
  if (identityKind === "git_repository") {
    const taskRepository = canonicalGitIdentity(
      workspaceIdentity,
      "workspace_identity",
    );
    if (
      !taskRepository ||
      (workspaceRevisionDigest !== null &&
        !GIT_REVISION_DIGEST_PATTERN.test(workspaceRevisionDigest)) ||
      (workspaceKind !== "canonical_checkout" &&
        workspaceKind !== "independent_git_worktree")
    ) return null;
    return {
      schema_version: DELIVERY_WORKSPACE_SCHEMA_VERSION,
      workspace_identity: taskRepository,
      identity_kind: identityKind,
      task_repository: taskRepository,
      ...(workspaceRevisionDigest === null
        ? {}
        : { workspace_revision_digest: workspaceRevisionDigest.toLowerCase() }),
      repository_source: repositorySource,
      workspace_kind: workspaceKind,
      peer_independent_worktree_required: peerIndependentWorktreeRequired,
    };
  }

  const localIdentity = localGoalIdentity(workspaceIdentity, "workspace_identity");
  if (
    !localIdentity || workspaceRevisionDigest !== null ||
    workspaceKind !== "local_goal_workspace" ||
    peerIndependentWorktreeRequired
  ) return null;
  return {
    schema_version: DELIVERY_WORKSPACE_SCHEMA_VERSION,
    workspace_identity: localIdentity,
    identity_kind: identityKind,
    task_repository: null,
    repository_source: repositorySource,
    workspace_kind: workspaceKind,
    peer_independent_worktree_required: false,
  };
}

export function buildDeliveryWorkspaceSnapshot(
  value: unknown,
): DeliveryWorkspaceSnapshot | null {
  const candidate = requireJsonObject(value, "delivery workspace observation");
  const identityKind = requireStringLiteral(
    candidate.identity_kind,
    DELIVERY_WORKSPACE_IDENTITY_KINDS,
    "identity_kind",
  );
  const workspaceKind = requireStringLiteral(
    candidate.workspace_kind,
    DELIVERY_WORKSPACE_KINDS,
    "workspace_kind",
  );
  return snapshot(
    requireNonEmptyString(candidate.workspace_identity, "workspace_identity"),
    identityKind,
    optionalNonEmptyString(
      candidate.workspace_revision_digest,
      "workspace_revision_digest",
    ),
    requireNonEmptyString(candidate.repository_source, "repository_source"),
    workspaceKind,
    requireBoolean(
      candidate.peer_independent_worktree_required,
      "peer_independent_worktree_required",
    ),
  );
}

export function normalizeDeliveryWorkspaceSnapshot(
  value: unknown,
): DeliveryWorkspaceSnapshot | null {
  if (value === null || value === undefined) return null;
  const candidate = requireJsonObject(value, "delivery workspace snapshot");
  const schemaVersion = optionalNonEmptyString(
    candidate.schema_version,
    "schema_version",
  );
  if (schemaVersion === LEGACY_DELIVERY_WORKSPACE_SCHEMA_VERSION) {
    const taskRepository = canonicalGitIdentity(
      candidate.task_repository,
      "task_repository",
    );
    if (!taskRepository) return null;
    const workspaceKind = requireStringLiteral(
      candidate.workspace_kind,
      ["canonical_checkout", "independent_git_worktree"] as const,
      "workspace_kind",
    );
    return snapshot(
      taskRepository,
      "git_repository",
      null,
      requireNonEmptyString(candidate.repository_source, "repository_source"),
      workspaceKind,
      requireBoolean(
        candidate.peer_independent_worktree_required,
        "peer_independent_worktree_required",
      ),
    );
  }
  if (schemaVersion !== DELIVERY_WORKSPACE_SCHEMA_VERSION) return null;

  const identityKind = requireStringLiteral(
    candidate.identity_kind,
    DELIVERY_WORKSPACE_IDENTITY_KINDS,
    "identity_kind",
  );
  const workspaceKind = requireStringLiteral(
    candidate.workspace_kind,
    DELIVERY_WORKSPACE_KINDS,
    "workspace_kind",
  );
  const normalized = snapshot(
    requireNonEmptyString(candidate.workspace_identity, "workspace_identity"),
    identityKind,
    optionalNonEmptyString(
      candidate.workspace_revision_digest,
      "workspace_revision_digest",
    ),
    requireNonEmptyString(candidate.repository_source, "repository_source"),
    workspaceKind,
    requireBoolean(
      candidate.peer_independent_worktree_required,
      "peer_independent_worktree_required",
    ),
  );
  if (!normalized) return null;
  const declaredRepository = optionalNonEmptyString(
    candidate.task_repository,
    "task_repository",
  );
  if (
    declaredRepository !== null &&
    canonicalGitIdentity(declaredRepository, "task_repository") !==
      normalized.task_repository
  ) return null;
  return normalized;
}

export function evaluateDeliveryWorkspace(value: unknown): JsonObject {
  const request = requestObject(value);
  const selectedOperation = operation(request.operation);
  return {
    schema_version: DELIVERY_WORKSPACE_RESULT_SCHEMA,
    workspace: selectedOperation === "build"
      ? buildDeliveryWorkspaceSnapshot(request.observation)
      : normalizeDeliveryWorkspaceSnapshot(request.workspace),
  };
}
